### PR TITLE
Remove Response from handler argument and get WS/HTTP working on same port

### LINKED ISSUE #156: HTTP.WebSockets served from the same port
This is more of a feature request that I will try myself, but thought I'd go ahead and create an issue first.

Now, we have `HTTP.Servers.Server.listen` and `HTTP.WebSockets.listen`. It would be nice to have a `listen` that works with both HTTP and WS on the same port.

I think it is as simple as taking the guts from `Server.listen` and `WebSockets.listen` and creating a new `listen` that is dispatched on if given both an HTTP handler and a WS handler (or some other flag).

### LINKED ISSUE #158: Remove Response argument if not needed
A `HandlerFunction` is of the form `(req,res) -> res`.

If the handler does not depend on `res`, we could do something like `(req,Response()) -> res`, but that feels like a waste to construct an object `Response()` that isn't needed. Is the compiler smart enough to optimize that away or does it really construct a `Response()` object?

In Pages, I imported `handle` and made these changes to make it work in order to avoid an empty `Response()`

```
SCHEMES = HTTP.Handlers.SCHEMES
EMPTYVAL = HTTP.Handlers.EMPTYVAL
HTTP.handle(h::HTTP.HandlerFunction,req::HTTP.Request) = h.func(req)
function HTTP.handle(r::HTTP.Router, req::HTTP.Request)
    # get scheme, host, split path into strings and get Vals
    m = Val(Symbol(req.method))
    # get the url/path of the request
    uri = HTTP.URI(req.target)
    s = get(SCHEMES, uri.scheme, EMPTYVAL)
    h = Val(Symbol(uri.host))
    p = uri.path
    segments = split(p, '/'; keep=false)
    # dispatch to the most specific handler, given the path
    vals = (get(r.segments, s, EMPTYVAL) for s in segments)
    handler = r.func(m, s, h, vals...)
    # pass the request & response to the handler and return
    return HTTP.handle(handler, req)
end
```

If there is no penalty for doing `(req,Response()) -> res`, I can just do this, otherwise, would you be open for a PR of the form above?

### PR DESCRIPTION
**Pull request by EricForgy**
I wasn't able to run `Pkg.test` because `addprocs` is broken on my machine, but this works inside Pages.jl.

Closes #158 
Closes #156

### CODE REVIEW

**Comment by EricForgy** [comment]
Not ready yet. I need to modify the tests too. Working on it 😅

**Comment by EricForgy** [comment]
Removed the args from handler tests, but they appear in some server tests too. I just removed those and waiting for Travis / AppVeyor 🙏

**Comment by EricForgy** [comment]
https://travis-ci.org/JuliaWeb/HTTP.jl/jobs/331411996

https://ci.appveyor.com/project/quinnj/http-jl/build/job/soi8th7cg08t9t41

I think the PR is ok now 👍 🙏 

Looks like there might be a non-deterministic client test that failed, but it passed on the previous two tests and the AppVeyor test passed (on v0.6).

**Comment by quinnj** [comment]
This seems fine to me, but definitely needs some docs. @EricForgy, would you mind writing up some general docs on handlers, and mention that each Request has it's own Response as a field member? That would be awesome.

**Comment by EricForgy** [comment]
Would it be possible to merge this and make a new issue for docs? 😅 

My code relies on these changes and I don't have time to update docs right now. There is a new PR merged and I hope to stay on master if possible.

**Comment by EricForgy** [comment]
There might be an easier way to do this?

What if instead of removing `resp` from all the methods, which is likely to break someone's code who uses HTTP.jl, we just add a few strategic methods like

```
handle(h,req::HTTP.Request) = HTTP(h,req,req.response)
```

?

**Comment by samoconnor** [comment]
Yeah, I'd go for cleaning up first.
If enough people need backwards-compatibity wrappers, that can be another PR. Or, it might be easier to support people through the change. How many existing users are there of this API anyway?

**Comment by EricForgy** [comment]
In order to get WS/HTTP to work, I was using my Pages.jl package, but Pages.jl depends on the original PR here, so I had to just add WS/HTTP changes to this PR. Hope that is alright 😅

**Comment by EricForgy** [comment]
Tests passed on AppVeyor v0.6 now 👍 

One server test failed on Travis v0.6, but I'm guessing it is non-determinstic.

**Comment by EricForgy** [comment]
Tests are passing now on v0.6 for both Travis and AppVeyor 👍 

https://travis-ci.org/JuliaWeb/HTTP.jl/jobs/331892650

https://ci.appveyor.com/project/quinnj/http-jl/build/job/0k09r8afv6duf04b

Sorry I didn't notice the v0.6 tests were not passing before 😅

File: test/server.jl [hunk 1]
```diff
@@ -49,9 +49,8 @@ sleep(2)
 
 # echo response
 serverlog = HTTP.FIFOBuffer()
-server = HTTP.Servers.Server((req, rep) -> begin
-    rep.body = req.body
-    return rep
+server = HTTP.Servers.Server((req) -> begin
+    return HTTP.Response(200,req.body)
```
Comment (samoconnor) [inline test/server.jl]: should be
```julia
req.response.body = req.body
return req.response
```
?

File: test/server.jl [hunk 2]
```diff
@@ -173,7 +172,7 @@ put!(server.in, HTTP.Servers.KILL)
 
 # keep-alive vs. close: issue #81
 port += 1
-tsk = @async HTTP.Servers.serve(HTTP.Servers.Server((req, res) -> (res.body = "Hello\n"; res), STDOUT), ip"127.0.0.1", port)
+tsk = @async HTTP.Servers.serve(HTTP.Servers.Server((req) -> (HTTP.Response(200,"Hello\n")), STDOUT), ip"127.0.0.1", port)
```
Comment (samoconnor) [inline test/server.jl]: Should be `(req.response.body = "Hello\n"; req.response)` ?

File: src/Servers.jl [hunk 3]
```diff
@@ -478,7 +478,7 @@ end
 Execute Request processing function `f(::HTTP.Request) -> HTTP.Response`.
 """
 
-function handle_request(f::Function, http::Stream)
+function HTTP.handle(f::Function, http::Stream)
```
Comment (samoconnor) [inline src/Servers.jl]: I'd rather not change the name of `handle_request` in Servers.jl.

File: src/Servers.jl [hunk 4]
```diff
@@ -487,5 +487,6 @@ function handle_request(f::Function, http::Stream)
     return
 end
 
+HTTP.handle(hf::HandlerFunction, http::Stream) = HTTP.handle(hf.func,http)
```
Comment (samoconnor) [inline src/Servers.jl]: I think this should be in handlers.jl.

**Comment by samoconnor** [comment]
Re:
```
            handler = HTTP.gethandler(router,http.message)
            HTTP.handle(handler,http)
```
Why not do this ?
```
            HTTP.handle_request(http) do request::HTTP.Request
                HTTP.handle(router, request)
            end
```

**Comment by samoconnor** [comment]
Hi @EricForgy, I'm going to merge this and tweak a few things per my comments. Please open a new PR if you need any subsequent changes.

**Comment by samoconnor** [comment]
See fd31e706eb402cc3be31deee1b6b5649b46054c0

**Comment by sairus7** [comment]
> Yay! 🎉
> 
> WS/HTTP working on the same port! 😃 🚀
> 
> ```
> function start(p = 8000)
>     global port = p
>     HTTP.listen(ip"127.0.0.1",p) do http
>         if is_upgrade(http.message)
>             HTTP.WebSockets.upgrade(http) do ws
>                 while !eof(ws);
>                     print(String(readavailable(ws)))
>                 end
>             end
>         else
>             handler = HTTP.gethandler(router,http.message)
>             HTTP.handle(handler,http)
>         end
>     end
> end
> ```

An updated version for those who may find this post and want to use it with HTTP.jl v0.9:
```
using Sockets, HTTP, HTTP.WebSockets

function start_server(port = 8081)
    HTTP.listen(ip"127.0.0.1", port) do http
        if WebSockets.is_upgrade(http.message)
            WebSockets.upgrade(http) do ws
                while !eof(ws)
                    print(String(readavailable(ws)))
                end
            end
        else
            handler = HTTP.Handlers.gethandler(router, http.message)
            HTTP.handle(handler, http)
        end
    end
end
```

**Pull request merged**